File: nodejs-express-api/controllers/subjects.js
import { Router } from 'express';
import { body } from 'express-validator';
import validateFormData from '../helpers/validate_form.js';
import DB from '../models/db.js';


const router = Router();




/**
 * Route to list subjects records
 * @GET /subjects/index/{fieldname}/{fieldvalue}
 */
router.get(['/', '/index/:fieldname?/:fieldvalue?'], async (req, res) => {  
	try{
		const query = {};
		let queryFilters = [];
		let where = {};
		let replacements = {};
		let fieldName = req.params.fieldname;
		let fieldValue = req.params.fieldvalue;
		
		if (fieldName){
			queryFilters.push(DB.filterBy(fieldName, fieldValue));
		}
		let search = req.query.search;
		if(search){
			let searchFields = DB.Subjects.searchFields();
			where[DB.op.or] = searchFields;
			replacements.search = `%${search}%`;
		}
		
		if(queryFilters.length){
			where[DB.op.and] = queryFilters;
		}
		query.raw = true;
		query.where = where;
		query.replacements = replacements;
		query.order = DB.getOrderBy(req, 'id', 'desc');
		query.attributes = DB.Subjects.listFields();
		let page = parseInt(req.query.page) || 1;
		let limit = parseInt(req.query.limit) || 10;
		let result = await DB.Subjects.paginate(query, page, limit);
		return res.ok(result);
	}
	catch(err) {
		return res.serverError(err);
	}
});


/**
 * Route to view Subjects record
 * @GET /subjects/view/{recid}
 */
router.get('/view/:recid', async (req, res) => {
	try{
		const recid = req.params.recid || null;
		const query = {}
		const where = {}
		where['id'] = recid;
		query.raw = true;
		query.where = where;
		query.attributes = DB.Subjects.viewFields();
		let record = await DB.Subjects.findOne(query);
		if(!record){
			return res.notFound();
		}
		return res.ok(record);
	}
	catch(err){
		return res.serverError(err);
	}
});


/**
 * Route to insert Subjects record
 * @POST /subjects/add
 */
router.post('/add/', 
	[
		body('subject_name').not().isEmpty(),
		body('hod').not().isEmpty(),
	], validateFormData
, async function (req, res) {
	try{
		let modeldata = req.getValidFormData();
		
		// check if hod already exist.
		let hodCount = await DB.Subjects.count({ where:{ 'hod': modeldata.hod } });
		if(hodCount > 0){
			return res.badRequest(`${modeldata.hod} already exist.`);
		}
		
		//save Subjects record
		let record = await DB.Subjects.create(modeldata);
		//await record.reload(); //reload the record from database
		const recid =  record['id'];
		const newValues = JSON.stringify(record); 
		req.writeToAuditLog({ recid, oldValues: null, newValues });
		
		return res.ok(record);
	} catch(err){
		return res.serverError(err);
	}
});


/**
 * Route to get  Subjects record for edit
 * @GET /subjects/edit/{recid}
 */
router.get('/edit/:recid', async (req, res) => {
	try{
		const recid = req.params.recid;
		const query = {};
		const where = {};
		where['id'] = recid;
		query.raw = true;
		query.where = where;
		query.attributes = DB.Subjects.editFields();
		let record = await DB.Subjects.findOne(query);
		if(!record){
			return res.notFound();
		}
		return res.ok(record);
	}
	catch(err){
		return res.serverError(err);
	}
});


/**
 * Route to update  Subjects record
 * @POST /subjects/edit/{recid}
 */
router.post('/edit/:recid', 
	[
		body('subject_name').optional({nullable: true}).not().isEmpty(),
		body('hod').optional({nullable: true}).not().isEmpty(),
	], validateFormData
, async (req, res) => {
	try{
		const recid = req.params.recid;
		let modeldata = req.getValidFormData({ includeOptionals: true });
		
		// check if hod already exist.
		let hodCount = await DB.Subjects.count({where:{'hod': modeldata.hod, 'id': {[DB.op.ne]: recid} }});
		if(hodCount > 0){
			return res.badRequest(`${modeldata.hod} already exist.`);
		}
		const query = {};
		const where = {};
		where['id'] = recid;
		query.raw = true;
		query.where = where;
		query.attributes = DB.Subjects.editFields();
		let record = await DB.Subjects.findOne(query);
		if(!record){
			return res.notFound();
		}
		const oldValues = JSON.stringify(record); //for audit trail
		await DB.Subjects.update(modeldata, {where: where});
		record = await DB.Subjects.findOne(query);//for audit trail
		const newValues = JSON.stringify(record); 
		req.writeToAuditLog({ recid, oldValues, newValues });
		return res.ok(modeldata);
	}
	catch(err){
		return res.serverError(err);
	}
});


/**
 * Route to delete Subjects record by table primary key
 * Multi delete supported by recid separated by comma(,)
 * @GET /subjects/delete/{recid}
 */
router.get('/delete/:recid', async (req, res) => {
	try{
		const recid = (req.params.recid || '').split(',');
		const query = {};
		const where = {};
		where['id'] = recid;
		query.raw = true;
		query.where = where;
		let records = await DB.Subjects.findAll(query);
		records.forEach(async (record) => { 
			//perform action on each record before delete
			const oldValues = JSON.stringify(record); //for audit trail
			req.writeToAuditLog({ recid: record['id'], oldValues });
		});
		await DB.Subjects.destroy(query);
		return res.ok(recid);
	}
	catch(err){
		return res.serverError(err);
	}
});
export default router;

File: python-flask-api/app/controllers/subjects.py
from flask import Blueprint
from sqlalchemy import text
from app import db, config, jsonify, request
from ..helpers.http_errors import InternalServerError, BadRequest, ResourceNotFound
from ..helpers.utils import utils
from ..models.subjects import *


Subjects_blueprint = Blueprint('Subjects', __name__)


# list Subjects records
# @param fieldname: filter table records by a field
# @param fieldvalue:  filter value
# @request.args search: search records
# @request.args orderby: sort records by field name
# @request.args ordertype: sort type (asc|desc)
# @request.args page: current page number
# @request.args limit: maximum number of records to select
@Subjects_blueprint.route('/')
@Subjects_blueprint.route('/index/<fieldname>/<fieldvalue>')
def list(fieldname=None, fieldvalue=None):
    try:
        query = Subjects.query
        search = request.args.get('search')
        if search:
            query = query.filter(
                Subjects.subject_name.like(f'%{search}%') | 
                Subjects.hod.like(f'%{search}%') | 
                Subjects.id.like(f'%{search}%') 
            )
        
        # filter by dynamic field name
        if fieldname:
            field_filter = text(f'{fieldname} = :fieldvalue').params(fieldvalue=fieldvalue)
            query = query.filter(field_filter)
        
        orderby = request.args.get('orderby')
        ordertype = request.args.get('ordertype', 'desc')
        if orderby:
            query = query.order_by(text(f'{orderby} {ordertype}'))
        else:
            query = query.order_by(text(f'subjects.id {ordertype}'))
        
        # fields to select
        query = query.with_entities(
            Subjects.subject_name,
            Subjects.hod,
            Subjects.id
        )
        page = int(request.args.get('page', 1))
        limit = int(request.args.get('limit', config.MAX_RECORD_COUNT))
        offset = ((page-1) * limit)
        total_records = query.count()
        records = query.limit(limit).offset(offset).all()
        record_count = len(records)
        total_pages = round(total_records/limit)
        records = [(row._asdict()) for row in records]
        
        # response object
        response = dict(
            records=records, 
            totalRecords=total_records,
            recordCount=record_count,
            totalPages=total_pages
        )
        return jsonify(response)
    except Exception as ex:
        return InternalServerError(ex)


# Select table record by ID
@Subjects_blueprint.route('/view/<rec_id>')
def view(rec_id=None):
    try:
        query = Subjects.query
        query = query.filter(Subjects.id == rec_id)
        query = query.with_entities(
            Subjects.id,
            Subjects.subject_name,
            Subjects.hod
        )
        
        record = query.first()
        if not record: return ResourceNotFound()
         
        record = record._asdict()
        
        # return result as json
        return jsonify(record)
        
    except Exception as ex:
        return InternalServerError(ex)


# Save form data to the Subjects table
@Subjects_blueprint.route('/add', methods=['POST'])
def add():
    try:
        modeldata = request.body
        form = SubjectsAddForm(modeldata)
        errors = [] # list of validation errors
        
        # validate subjects form data
        if not form.validate():
            errors.append(form.errors)
        
        if errors:
            return BadRequest(errors)
        
        record = Subjects()
        form.populate_obj(record)

        # check if hod record already exist in the database
        rec_value = str(modeldata['hod'])
        rec_exist = utils.is_unique(Subjects, "hod", rec_value)
        if rec_exist:
            return BadRequest(rec_value + " Already exist!")
        
        # save subjects records
        db.session.add(record)
        db.session.commit()
        db.session.flush()
        rec_id = record.id
         
        record = record._asdict()
        return jsonify(record)
    except Exception as ex:
        return InternalServerError(ex)


# Select record by table primary key and update with form data
@Subjects_blueprint.route('/edit/<rec_id>', methods=['GET', 'POST'])
def edit(rec_id=None):
    try:
        query = Subjects.query
        query = query.filter(Subjects.id == rec_id)
        record = query.first()
        if not record: return ResourceNotFound()
        
        if request.method == 'POST':
            errors = []
            modeldata = request.body
            form = SubjectsEditForm(modeldata, obj=record)
            
            if not form.validate():
                errors.append(form.errors)
            # check if hod already exist in the database
            rec_value = str(modeldata['hod'])
            rec_exist = utils.is_unique(Subjects, "hod", rec_value, "id", rec_id)
            if rec_exist:
                return BadRequest(rec_value  + " Already exist!")
            
            if errors:
                return BadRequest(errors)
            
            # save Subjects record
            form.populate_obj(record)
            db.session.commit()
         
        record = record._asdict()
        return jsonify(record)
    except Exception as ex:
        return InternalServerError(ex)


# Delete record from the database
# Support multi delete by separating record id by comma.
@Subjects_blueprint.route('/delete/<rec_id>')
def delete(rec_id):
    query = Subjects.query
    arr_id = rec_id.split(',')
    try:
        query = query.filter(Subjects.id.in_(arr_id))
        query.delete(synchronize_session=False)
        db.session.commit()
        
        return jsonify(arr_id)
    except Exception as ex:
        return InternalServerError(ex)
